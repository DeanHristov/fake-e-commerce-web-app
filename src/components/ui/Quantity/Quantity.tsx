import { FC, useCallback, useState } from 'react';

export interface IQuantityProps {
  onIncrease: (value: number) => void;
  onDecrease: (value: number) => void;
  initialValue?: number;
  maxValue?: number;
  className?: string;
}

const Quantity: FC<IQuantityProps> = ({
  initialValue,
  onIncrease,
  onDecrease,
  className,
  maxValue = Infinity,
}) => {
  const [quantity, setQuantity] = useState<number>(initialValue || 1);

  const handlerOnIncrease = useCallback(() => {
    const nextValue: number = quantity + 1;

    if (nextValue < maxValue) {
      onIncrease(nextValue);
      setQuantity(nextValue);
    }
  }, [quantity, onIncrease]);

  const handlerOnDecrease = useCallback(() => {
    const nextValue: number = quantity - 1;

    if (nextValue > 0) {
      onDecrease(nextValue);
      setQuantity(nextValue);
    }
  }, [quantity, onDecrease]);

  return (
    <div
      className={`flex justify-center items-center bg-gray-100 rounded-md border border-gray-400 ${className}`}
    >
      <button
        className="text-lg w-20 py-1 text-red-400 font-medium border-r border-r-gray-400"
        onClick={handlerOnDecrease}
      >
        {'-'}
      </button>
      <span className="w-full px-1 py-1 text-center">{quantity}</span>
      <button
        className="text-lg w-20 py-1 text-green-400 font-medium border-l border-l-gray-400"
        onClick={handlerOnIncrease}
      >
        {'+'}
      </button>
    </div>
  );
};

export default Quantity;
