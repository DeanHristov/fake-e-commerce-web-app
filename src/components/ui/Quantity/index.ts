import Quantity, { IQuantityProps } from './Quantity';

export type { IQuantityProps };
export default Quantity;
