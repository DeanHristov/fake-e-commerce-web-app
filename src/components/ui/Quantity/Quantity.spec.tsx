import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Quantity from './';

describe('UI/Quantity <Quantity {...} />', () => {
  it('Should render a default state of the Component', () => {
    const { container } = render(
      <Quantity onDecrease={jest.fn} onIncrease={jest.fn} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('Should be able to see the initial value', () => {
    const { getByText } = render(
      <Quantity onIncrease={jest.fn} onDecrease={jest.fn} initialValue={1} />,
    );

    expect(getByText(1)).toBeVisible();
  });

  it('Should be able to increase the initial value', async () => {
    const mockCallback = jest.fn();
    const { getAllByRole } = render(
      <Quantity
        onIncrease={mockCallback}
        onDecrease={jest.fn}
        initialValue={1}
      />,
    );

    const $buttons: HTMLElement[] = getAllByRole('button');

    await userEvent.click($buttons[1]);

    expect(mockCallback).toHaveBeenCalledWith(2);
  });

  it('Should be able to decrease the initial value', async () => {
    const mockCallback = jest.fn();
    const { getAllByRole } = render(
      <Quantity
        onIncrease={jest.fn}
        onDecrease={mockCallback}
        initialValue={2}
      />,
    );

    const $buttons: HTMLElement[] = getAllByRole('button');

    await userEvent.click($buttons[0]);

    expect(mockCallback).toHaveBeenCalledWith(1);
  });

  it('Should not be able to exceeded the max value', async () => {
    const mockCallback = jest.fn();
    const { getAllByRole } = render(
      <Quantity
        onIncrease={mockCallback}
        onDecrease={jest.fn}
        initialValue={1}
        maxValue={1}
      />,
    );

    const $buttons: HTMLElement[] = getAllByRole('button');

    await userEvent.click($buttons[1]);

    expect(mockCallback).not.toHaveBeenCalled();
  });

  it('Should not be able to exceeded a negative value', async () => {
    const mockCallback = jest.fn();
    const { getAllByRole } = render(
      <Quantity
        onIncrease={mockCallback}
        onDecrease={jest.fn}
        initialValue={1}
      />,
    );

    const $buttons: HTMLElement[] = getAllByRole('button');

    await userEvent.click($buttons[0]);

    expect(mockCallback).not.toHaveBeenCalled();
  });
});
