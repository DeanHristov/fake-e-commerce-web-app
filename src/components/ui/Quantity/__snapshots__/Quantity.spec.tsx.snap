// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI/Quantity <Quantity {...} /> Should render a default state of the Component 1`] = `
<div>
  <div
    class="flex justify-center items-center bg-gray-100 rounded-md border border-gray-400 undefined"
  >
    <button
      class="text-xl w-20 text-red-400 font-medium border-r border-r-gray-400"
    >
      -
    </button>
    <span
      class="w-full px-1  text-center"
    >
      1
    </span>
    <button
      class="text-xl w-20  text-green-400 font-medium border-l border-l-gray-400"
    >
      +
    </button>
  </div>
</div>
`;
